x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlEntityDescriptorBuilder.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.authc.saml;
+
+import org.elasticsearch.cli.ExitCodes;
+import org.elasticsearch.cli.Terminal;
+import org.elasticsearch.cli.UserException;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.xpack.core.security.authc.RealmConfig;
+import org.elasticsearch.xpack.core.security.authc.RealmSettings;
+import org.elasticsearch.xpack.core.security.authc.saml.SamlRealmSettings;
+import org.opensaml.saml.saml2.core.AuthnRequest;
+import org.opensaml.saml.saml2.metadata.EntityDescriptor;
+
+import java.util.Locale;
+import java.util.Map;
+
+public class SamlEntityDescriptorBuilder {
+    private final EntityDescriptor entityDescriptor;
+
+    public EntityDescriptor getEntityDescriptor() {
+        return entityDescriptor;
+    }
+
+    public SamlEntityDescriptorBuilder(SamlRealm samlRealm) throws Exception{
```

Comment:
So I had missed we already have `SamlSpMetadataBuilder`. This is great because we could use that and not introduce SamlEntityDescriptorBuilder`. You already have everything you need there to use in the Transport action I think. 

To be more concrete, `buildEntityDescriptorFromSamlRealm` that you already had in SamlMetadataCommand, could live in the transport action ( as a method of its own or as part of `prepareMetadata` ) and make use of SamlSpMetadataBuilder

Reply:
Refactored